tests/sync/session/wait_for_completion.cpp

```diff
@@ -0,0 +1,256 @@
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "catch.hpp"
+
+#include "sync/session/session_util.hpp"
+
+#include "util/event_loop.hpp"
+
+#include <realm/util/scope_exit.hpp>
+
+using namespace realm;
+using namespace realm::util;
+
+TEST_CASE("SyncSession: wait_for_download_completion() API", "[sync]") {
+    if (!EventLoop::has_implementation())
+        return;
+
+    auto cleanup = util::make_scope_exit([=]() noexcept { SyncManager::shared().reset_for_testing(); });
+    SyncServer server;
+    // Disable file-related functionality and metadata functionality for testing purposes.
+    SyncManager::shared().configure_file_system(tmp_dir(), SyncManager::MetadataMode::NoMetadata);
+
+    std::atomic<bool> handler_called(false);
+
+    SECTION("works properly when called after the session is bound") {
+        auto user = SyncManager::shared().get_user("user-async-wait-download-1", "not_a_real_token");
+        auto session = sync_session(server, user, "/async-wait-download-1",
+                                    [](const std::string&, const std::string&) { return s_test_token; },
+                                    [](auto, auto) { });
+        EventLoop::main().run_until([&] { return sessions_are_active(*session); });
+        // Register the download-completion notification
+        CHECK(session->wait_for_download_completion([&](auto) {
+            handler_called = true;
+        }));
+        EventLoop::main().run_until([&] { return handler_called == true; });
+    }
+
+    SECTION("works properly when called on a session waiting for its access token") {
+        auto user = SyncManager::shared().get_user("user-async-wait-download-2", "not_a_real_token");
+        std::atomic<bool> login_handler_called(false);
+        auto server_path = "/async-wait-download-2";
+        std::shared_ptr<SyncSession> session = sync_session_with_bind_handler(server, user, server_path,
+                                                                              [&](auto, auto, std::shared_ptr<SyncSession> s){
+                                                                                  session = std::move(s);
+                                                                                  login_handler_called = true;
+                                                                              },
+                                                                              [](auto, auto) { });
+        // Register the download-completion notification
+        REQUIRE(session->wait_for_download_completion([&](auto) {
+            handler_called = true;
+        }));
+        EventLoop::main().run_until([&] { return login_handler_called == true; });
+        REQUIRE(session);
+        REQUIRE(handler_called == false);
+        // Spin the loop many times.
+        std::atomic<int> spin_count(0);
+        EventLoop::main().run_until([&] { spin_count++; return spin_count > 100; });
```

Comment:
This single section of the test appears to run for over a second, predominantly due to this spinning. This pattern is repeated a few times in these new tests, adding around 5 seconds to the total runtime of all object store tests. That doubles how long it takes the object store tests to run. Is it necessary to spin for so long?